src/diffusers/loaders/unet.py

```diff
@@ -50,6 +51,10 @@
 LORA_WEIGHT_NAME = "pytorch_lora_weights.bin"
 LORA_WEIGHT_NAME_SAFE = "pytorch_lora_weights.safetensors"
 
+PEFT_WEIGHT_NAME = "adapter_model.bin"
+PEFT_WEIGHT_NAME_SAFE = "adapter_model.safetensors"
+PEFT_CONFIG_NAME = "adapter_config.json"
```

Comment:
```suggestion
PEFT_WEIGHT_NAME = "adapter_model.bin"
PEFT_WEIGHT_NAME_SAFE = "adapter_model.safetensors"
PEFT_CONFIG_NAME = "adapter_config.json"
```


@BenjaminBossan @pacman100 @younesbelkada @apolinario  - We need to discuss this a bit. I'm all in favor of supporting `adapter_config.json` mid-term, but we have some peculiarities here:
- Currently LoRAs serialization format is a single file that contains weights for both the Unet and the text_encoder. However, it's also not uncommon that one only trains the unet in which case the single file only contains the unet
- I'd say pretty much always text encoder and unet LoRAs are trained together, but I do see use cases where one might want to load two different LoRA files (one for it's text encoder, one for it's unet)

=> As a consequence the main "loading" function of LoRA is part of the pipeline:

```py
pipeline = DiffusionPipeline.from_pretrained("...")
pipeline.load_lora_weights("...")  # this internally then dispatches to loading the text encoder specific loras to the text encoder and the unet specific loras to the unet
```

**However**, it's also very important that we allow loading LoRAs just from the unet or the text encoder:

```py
pipeline.text_encoder.load_lora   # <- here we call Transformers' lora loading function
pipeline.unet.load_lora  # <- this needs to be implemented well in Diffusers
```

So the question here is, what format and serialization should we use for PEFT that is both:
- Easy to use for training (meaning easy to use **just** on the `unet` object when no pipeline is loaded)
- Easy to use for inference (when loaded via `pipeline.load_lora_weights`)
- Easy to use to mix different LoRAs

Can we allow `adapter.json` to have a config for either just a text encoder, just a unet, both text encoder and unet
Can `adapter.safetensors` have both formats?

Reply:
Not sure if I grasped the problem completely, but the 2nd part (separate LoRAs for text encoder and unet) would be the easier one, right? We would just point to different folders, each with their own `adapter_config.json` etc.

For the combined adapter, we would need to find a new solution, it could be some kind of convention (sub-folder names) or we would need to have a single adapter that takes care of both, but that seems unwieldy and more error prone.

Another use case I'd like to bring up, not sure if it's relevant at this moment: Having a pipeline load multiple different LoRA adapters (or even other type of adapters).

Reply:
> but the 2nd part (separate LoRAs for text encoder and unet) would be the easier one, right? 

Yes, it'd def be cleaner as a serialization format, but it's very much not the convention right now. At the moment, all LoRAs always only have a single safetensors file. So if we introduce a new 4 file convention (text_encoder's adapter_config.json, text_encoder's adapter_model.safetensors unet's adapter_config.json and unet's adapter_model.safetensors) it might be quite confusing for the community. I'm not too happy about creating a new folder structure as this wasn't super well perceived in the first place.

Reply:
I prefer the file format with subfolders for unet/text-encoder each having a `adapter_model.safetensors` and `adapter_config.json`. This is exactly micing the folder structure of full finetuning/pretrained SD/SDXL in Diffusers just that the config/weights are only for adapters. So, that would be clearly following the standard persistence format of Diffusers

Reply:
For portability, I had written a conversion script which would put both the unet and text-encoder weights and their config's in a single `safetensors` file. The file is this: https://github.com/pacman100/peft-dreambooth/blob/master/peft_utils/merge_peft_sd_ckpt_to_single_safetensors.py

Reply:
But, I believe the following structure if pretty flexible and inline with the original storage standard of Diffusers itself.
![Screenshot 2023-11-30 at 10 17 38 AM](https://github.com/huggingface/diffusers/assets/13534540/fefbf922-e8aa-4646-96b4-a7c328107bd9)

Reply:
It looks infeasible to support multiple configs inside `adapter_config.json` as this requirement is very Diffusers specific and largely deviating from the standard of storing the model and related config in a single repo.

Reply:
I think we can have Diffusers specific utilities either in PEFT or Diffusers which saves everything in a single `safetensors` file using a similar conversion method to the one above which I use. So, when diffusers code calls below function, it would use those PEFT utils when PEFT is enabled. A few more utils for loading the single safetensors handling cases when only unet is present, only text encoder is present or both are present which should be straightforward.
```
StableDiffusionXLPipeline.save_lora_weights(
            save_directory=args.output_dir,
            unet_lora_layers=unet_lora_layers,
            text_encoder_lora_layers=text_encoder_lora_layers,
            text_encoder_2_lora_layers=text_encoder_2_lora_layers,
        )
```

Reply:
This way, current paradigm of single safetensors for LoRA would be achieved for PEFT and users are happy. But, this single safetensors file won't work in Auto1111 as it requires changing the state dict keys which is also implemented here https://github.com/pacman100/peft-sd-webui-additional-networks/blob/main/scripts/peft_lora.py

Reply:
Thanks for your thoughts. But the thing is the diffusion community very much prefers the single file-format at least for LoRAs.

Reply:
then these comments https://github.com/huggingface/diffusers/pull/5958/files#r1410149384 and https://github.com/huggingface/diffusers/pull/5958/files#r1410163690 would be the recommended way to have Diffusers specific utils.

Reply:
Another thing to keep in mind is that the utils need to generic. For example, A PR to add OFT in PEFT is close to being merged. It follows the standard format of subfolders with adapter+config files. The utils need to convert both LoRA and OFT into single safetensors format.

Reply:
Having a utility to output a single file seems to be the best choice here without compromising too much.